src/charm.py

```diff
@@ -589,21 +593,36 @@ def _update_cert(self):
                 self.server_cert.ca_cert,  # pyright: ignore
                 make_dirs=True,
             )
+        except ConnectionError:
+            logger.error("Could not push cert files. Pebble refused connection. Shutting down?")
+            return
 
-            # Repeat for the charm container. We need it there for loki client requests.
-            # (and charm tracing and logging TLS)
-            ca_cert_path.parent.mkdir(exist_ok=True, parents=True)
-            ca_cert_path.write_text(self.server_cert.ca_cert)  # pyright: ignore
-        else:
+        # Repeat for the charm container. We need it there for loki client requests.
+        # (and charm tracing and logging TLS)
+        ca_cert_path.parent.mkdir(exist_ok=True, parents=True)
+        ca_cert_path.write_text(self.server_cert.ca_cert)  # pyright: ignore
+
+    def _remove_cert_files(self):
+        ca_cert_path = Path(self._ca_cert_path)
+        try:
             self._loki_container.remove_path(CERT_FILE, recursive=True)
             self._loki_container.remove_path(KEY_FILE, recursive=True)
             self._loki_container.remove_path(ca_cert_path, recursive=True)
+        except ConnectionError:
+            logger.error("Could not remove cert files. Pebble refused connection. Shutting down?")
+            return
 
-            # Repeat for the charm container.
-            ca_cert_path.unlink(missing_ok=True)
+        # Repeat for the charm container.
+        ca_cert_path.unlink(missing_ok=True)
 
-        self._loki_container.exec(["update-ca-certificates", "--fresh"]).wait()
-        subprocess.run(["update-ca-certificates", "--fresh"])
+    def _update_ca_certificates(self):
+        try:
+            self._loki_container.exec(["update-ca-certificates", "--fresh"]).wait()
+            subprocess.run(["update-ca-certificates", "--fresh"])
+        except (ConnectionError, ExecError) as e:
+            logger.error("Could not run update-ca-certificates in workload container. %s", e)
+        except subprocess.SubprocessError as e:
+            logger.error("Could not run update-ca-certificates in charm container. %s", e)
```

Comment:
This feels like we are just sweeping errors under the rug. Isn't it a problem if the `update-ca-certificates` command fails?

Reply:
You are right.

WDYT about this?

```diff
@@ -84,6 +84,7 @@ class CompositeStatus(TypedDict):
     config: Tuple[str, str]
     rules: Tuple[str, str]
     retention: Tuple[str, str]
+    certificates: Tuple[str, str]
 
 
 def to_tuple(status: StatusBase) -> Tuple[str, str]:
@@ -132,6 +133,7 @@ class LokiOperatorCharm(CharmBase):
                 config=to_tuple(ActiveStatus()),
                 rules=to_tuple(ActiveStatus()),
                 retention=to_tuple(ActiveStatus()),
+                certificates=to_tuple(ActiveStatus()),
             ),
             fresh_install=True,
         )
@@ -620,9 +622,13 @@ class LokiOperatorCharm(CharmBase):
             self._loki_container.exec(["update-ca-certificates", "--fresh"]).wait()
             subprocess.run(["update-ca-certificates", "--fresh"])
         except (ConnectionError, ExecError) as e:
-            logger.error("Could not run update-ca-certificates in workload container. %s", e)
+            msg = f"Could not run update-ca-certificates in workload container. {e}"
+            logger.error(msg)
+            self._stored.status["certificates"] = to_tuple(BlockedStatus(msg))
         except subprocess.SubprocessError as e:
-            logger.error("Could not run update-ca-certificates in charm container. %s", e)
+            msg = f"Could not run update-ca-certificates in charm container. {e}"
+            logger.error(msg)
+            self._stored.status["certificates"] = to_tuple(BlockedStatus(msg))
```

Reply:
I think this needs to be error status. Blocked status implies that the user should can fix it. We should just raise the error or raise a custom error like `CACertsError` or something. The bonus of making it an error status is Juju will retry the hook automatically and it might resolve itself.

Reply:
You are right about not setting Blocked status....  raising the error and letting the charm go into error state maybe is a good idea here.
Besides I thought about using other two options: `tenacity` or `.defer()`...

What do you think @sed-i @dstathis ?

Reply:
I think tenacity would likely be better. Retry for a few minutes, then raise an exception if that fails.

Reply:
This PR started because of a missing `can_connect()` guard in the lib, the changes in `charm.py` seemed trivial, but after this exchange of messages it is clear that we should think better about it. 

Let's follow this conversation in a separate issue: https://github.com/canonical/loki-k8s-operator/issues/434 and merge only the changes to the lib